_posts/2023-03-27-group-equivariant-cnn-2/group-equivariant-cnn-2.Rmd

```diff
@@ -0,0 +1,374 @@
+```
+
+Convolutional neural networks (CNNs) are great -- they're able to detect features in an image no matter where. Well, not exactly. They're not indifferent to just any kind of movement. Shifting up or down, or left or right, is fine; rotating around an axis is not. That's because of how convolution works: traverse by row, then traverse by column (or the other way round). If we want "more" (e.g., successful detection of an upside-down object), we need to extend convolution to an operation that is *rotation-equivariant*. An operation that is *equivariant* to some type of action will not just register the moved feature per se, but also, keep track of which concrete action made it appear where it is.
+
+**This is the second post in a series that introduces group-equivariant CNNs (GCNNs)***.* The [first](https://blogs.rstudio.com/ai/posts/2023-03-09-group-equivariant-cnn-1/) was a high-level introduction to why we'd want them, and how they work. There, we introduced the key player, the symmetry group that specifies which kinds of transformations are to be treated equivariantly. If you haven't, please take a look at that post first, since here I'll make use of terminology and concepts it introduced.
+
+Today, we code a simple GCNN from scratch. Code and presentation tightly follow a [notebook](https://github.com/phlippe/uvadlc_notebooks/blob/master/docs/tutorial_notebooks/DL2/Geometric_deep_learning/tutorial1_regular_group_convolutions.ipynb) provided as part of University of Amsterdam's 2022 [Deep Learning Course](https://uvadl2c.github.io/). They can't be thanked enough for making available such excellent learning materials.
+
+In what follows, my intent is to explain the general thinking, and how the resulting architecture is built up from smaller modules, assigned a clear purpose each. For that reason, I won't reproduce all the code here; instead, I'll make use of the package [`gcnn`](https://github.com/skeydan/gcnn)`.` Its methods are heavily annotated; so to see some details, don't hesitate to look at the code.
+
+As of today, `gcnn` implements one symmetry group: $C_4$, the one that has been serving as a running example throughout post one. It is straightforwardly extensible, though, making use of class hierarchies throughout.
+
+## Step 1: The symmetry group $C_4$
+
+In coding a GCNN, the first thing we need to provide is an implementation of the symmetry group we'd like to use. Here, it is $C_4$, the four-element group that rotates by 90 degrees.
+
+We can ask `gcnn` to create one for us, and inspect its elements.
+
+```{r}
+# remotes::install_github("skeydan/gcnn")
+library(gcnn)
+library(torch)
+
+C_4 <- CyclicGroup(order = 4)
+elems <- C_4$elements()
+elems
+```
+
+    torch_tensor
+     0.0000
+     1.5708
+     3.1416
+     4.7124
+    [ CPUFloatType{4} ]
+
+Elements are straightforwardly represented by their respective rotation angles: $0$, $\frac{\pi}{2}$, $\pi$, and $\frac{3 \pi}{2}$.
```

Comment:
```suggestion
Elements are represented by their respective rotation angles: $0$, $\frac{\pi}{2}$, $\pi$, and $\frac{3 \pi}{2}$.
```

Reply:
"Straightforward" (or it's mouthful of an adverb-form, straightforwardly) is one of those words like clearly, simply, obviously, just, etc. that I think can be cut much of the time. (If it's straightforward, the reader will garner that…otherwise, they feel bad for finding it non-obvious).

Reply:
Hehe :-) Caught me again :-)
Didn't want to write "simply", but wanted to contrast this with the method that's discussed next (which is less "straightforward") ... Still, you're right, it's totally enough if I say "less straightforward" when I get there!